util/file_util.py

```diff
@@ -0,0 +1,922 @@
+                self._fd = storage.fileio.BlobReader(blob)
+            else:
+                blob = file_get_gcs_blob(filename, exists=False)
+                self._fd = storage.fileio.BlobWriter(blob)
+        else:
+            logging.debug(f'Opening file:{filename} mode:{self._mode}')
+            self._fd = open(filename,
+                            mode=self._mode,
+                            encoding=self._encoding,
+                            newline=self._newline)
+
+    def __del__(self):
+        '''Cleanup any temporary files.'''
+        self.__exit__(None, None, None)
+
+    def __enter__(self):
+        '''Return the file handle for the local file.'''
+        return self._fd.__enter__()
+
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        '''Cleanup after file IO is complete.
+        Close the file handles and copy temporary files and delete them.
+        '''
+
+        # Close the file handle.
+        if self._fd:
+            logging.debug(
+                f'Closing file:{self._filename}, tmp:{self._tmp_filename}')
+            self._fd.__exit__(exc_type, exc_value, exc_tb)
+            self._fd = None
+
+        # Copy over the temp file written into the original file.
+        if self._tmp_filename and self._mode.startswith('w'):
+            if file_is_gcs(self._filename):
+                # Copy the tmp file to the requested GCS file.
+                logging.debug(
+                    f'Copying {self._tmp_filename} to {self._filename}')
+                with open(self._tmp_filename, mode='rb') as tmpfile:
+                    blob = file_get_gcs_blob(self._filename, exists=False)
+                    gcsfile = storage.fileio.BlobWriter(blob)
+                    _copy_file_chunks(tmpfile, gcsfile)
+                    gcsfile.close()
+            elif file_is_google_spreadsheet(self._filename):
+                file_copy_to_spreadsheet(self._tmp_filename, self._filename)
+            else:
+                # Rename the tmp file to the original local file.
+                if self._tmp_filename:
+                    logging.debug(
+                        f'Renaming {self._tmp_filename} to {self._filename}')
+                    os.rename(self._tmp_filename, self._filename)
+                self._tmp_filename = None
+
+        # Delete the temporary file
+        if self._tmp_filename:
+            logging.debug(f'Deleting tempfile: {self._tmp_filename}')
+            os.remove(self._tmp_filename)
+            self._tmp_filename = None
+
+    def read(self, *args, **kwargs):
+        '''Read from the file handle.'''
+        return self._fd.read(*args, **kwargs)
+
+    def readline(self, *args, **kwargs):
+        '''Read from the file handle.'''
+        return self._fd.readline(*args, **kwargs)
+
+    def write(self, *args, **kwargs):
+        '''Write into the file handle.'''
+        return self._fd.write(*args, **kwargs)
+
+
+# Utilities for files.
```

Comment:
nit: Are the below exported from the API?  if not use `_` prefix?

Reply:
The functions below are used by the events and statvar pipelines.